### ISSUE TITLE
maxage not working with $http and localstorage with passive deleteOnExpire setting 

### ISSUE THREAD

**Issue by ekallevig** [jmdobry/angular-cache#224]
I'm using angular-cache with $http and localStorage. I use a maxAge and "passive" setting for deleteOnExpire.  The caching works fine, but after exceeding the maxAge time (expired), it continues to fetch from cache. If I change deleteOnExpire to 'aggressive', it works as expected (fetches from server after maxAge has elapsed). When I switch it back to memory storageMode (with passive), it works as expected (after item expires, it fetches from server).

My reading of the docs was that passive simply means it doesn't clear the expired value until the cache is requested again (vs aggressive, where it clears it the second the maxAge has elapsed).

Am I doing something wrong or is this expected?

**Comment by rsc092020** [comment]
+1. I have seen the same results, where when you use localStorage (or sessionStorage) with $http and passive deleteOnExpire and maxAge, the cache items appear to never expire.

**Comment by smutnyleszek** [comment]
Same here, also with `sessionStorage` as mentioned above.

**Comment by jmdobry** [comment]
So the settings to reproduce are:

```
{
  storageMode: 'localStorage',
  deleteOnExpire: 'passive'
}
```

plus using $http, correct?

**Comment by rsc092020** [comment]
Yes, that's correct, plus a maxAge. It will just stay around in localStorage without getting fetched again from server, even after the expiration according to maxAge.

**Comment by robations** [comment]
I think this is related. I'm using (was using) this config with the `$http` service:

    {
        maxAge: 4 * 60 * 60 * 1000,
        deleteOnExpire: "passive",
        storageMode: "localStorage",
    }

Although I could see keys being added into local storage, every time I refreshed the page the cache would be ignored. Changing "passive" to "aggressive" is an okay workaround for me. If I can spend any time looking at this later I will.

**Comment by jmdobry** [comment]
A plunker to reproduce would be great, if you can

**Comment by robations** [comment]
Probably need to open this in a separate window otherwise you'll be hit by iframe restrictions on localstorage. Hit the update button and it will make a request that should be cached for 5 minutes. (Open the console to see network and log feedback.)

As far as I can see, this example is working correctly, so something else must be triggering the issue.

https://plnkr.co/edit/kApzDm20xH7rA2eeKxku?p=preview

Need to move on but will investigate later.

**Comment by rsc092020** [comment]
The plunkr by @robations is an example of the original issue that was brought up. After fetching the first time, it will just consistently use the cache after that even if it's older than the max-age (Try lowering the max-age to make this easier to see). I haven't had enough time to look at this close enough to come up with a fix, but basically what is happening is it will hit this [code](https://github.com/jmdobry/angular-cache/blob/master/dist/angular-cache.js#L325-L343) in the `get` function when trying to retrieve a cached http request that is in localStorage (or anything that isn't in-memory):

    if ($$storage) {
      if ($$promises[key]) {
        return $$promises[key];
      }

      var itemJson = $$storage().getItem(this.$$prefix + '.data.' + key);

      if (itemJson) {
        item = utils.fromJson(itemJson);
      } else {
        return;
      }
    } else if (utils.isObject($$data)) {
      if (!(key in $$data)) {
        return;
      }

      item = $$data[key];
    }


then down below is this piece of [code](https://github.com/jmdobry/angular-cache/blob/master/dist/angular-cache.js#L364-L365) that will remove the item if it's expired:

    if (this.$$deleteOnExpire === 'passive' && 'expires' in item && item.expires < now) {
        this.remove(key);

But it will return before it reaches this code in certain conditions, specifically in this case, when using a storageMode that is not memory and when the item being cached is a promise. There could be other cases where this happens too, though.

**Comment by delagen** [comment]
I think
```
 const getHandler = (shouldStore, isError) => {
      return (v) => {
        if (shouldStore) {
          this.$$promises[key] = undefined
```
should be rewritten with
```
 const getHandler = (shouldStore, isError) => {
      return (v) => {
        this.$$promises[key] = undefined
        if (shouldStore) {
```

so in next cache.get it reach cleanup part

**Comment by clarkieryan** [comment]
@jmdobry  I'm having a  similar problem - looks as if you have fixed this on the CacheFactory repo - however not updated the dependency on this repository?
